### PR TITLE
Vpc machine reconciliation

### PR DESCRIPTION
**Pull request by cjschaef**
**What this PR does / why we need it**:
Support additional options and utilize the extended VPC Infrastructure (Network Status) options for VPC Machine reconciliation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #  https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/1977

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Extend support options for reconciliation of VPC Machines
```

### CODE REVIEW

File: cloud/scope/machine.go [hunk 1]
```diff
@@ -127,37 +155,72 @@ func (m *MachineScope) CreateMachine() (*vpcv1.Instance, error) {
...
+		subnetDetails, err := m.IBMVPCClient.GetVPCSubnetByName(m.IBMVPCMachine.Spec.PrimaryNetworkInterface.Subnet)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving subnet ID for machine %s: %w", m.IBMVPCMachine.Name, err)
+		} else if subnetDetails != nil {
+			subnetIdentity.ID = subnetDetails.ID
+		} else {
+			subnetIdentity.ID = &m.IBMVPCMachine.Spec.PrimaryNetworkInterface.Subnet
+		}
+	}
+	primaryNetworkInterface := &vpcv1.NetworkInterfacePrototype{
```
Comment (cjschaef) [inline cloud/scope/machine.go:186]: If there is time, I may try to update this PR with the logic to perform SecurityGroup lookups (from name to ID) to complete the network interface configuration.

Comment (cjschaef) [reply cloud/scope/machine.go:185]: Added, I will attempt to do some basic validation testing and see if I can also add one or more unit tests as well.

File: cloud/scope/machine.go [hunk 2]
```diff
@@ -127,37 +155,72 @@ func (m *MachineScope) CreateMachine() (*vpcv1.Instance, error) {
...
+	// Build common field resources, as unique InstancePrototype's are defined based on machine source.
+	profile := &vpcv1.InstanceProfileIdentity{
+		Name: &m.IBMVPCMachine.Spec.Profile,
+	}
+
+	subnetIdentity := &vpcv1.SubnetIdentity{}
+	// Find subnet ID from Cluster Network Status, if possible.
+	// If Network Status is available, attempt to retrieve subnet ID from there.
+	if m.IBMVPCCluster.Status.Network != nil {
+		if subnet, ok := m.IBMVPCCluster.Status.Network.ControlPlaneSubnets[m.IBMVPCMachine.Spec.PrimaryNetworkInterface.Subnet]; ok {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply cloud/scope/machine.go:167]: I'll add a check.

File: cloud/scope/machine.go [hunk 3]
```diff
@@ -127,37 +155,72 @@ func (m *MachineScope) CreateMachine() (*vpcv1.Instance, error) {
...
 	}
 
-	imageID, err := fetchImageID(m.IBMVPCMachine.Spec.Image, m)
-	if err != nil {
-		record.Warnf(m.IBMVPCMachine, "FailedRetriveImage", "Failed image retrival - %v", err)
-		return nil, fmt.Errorf("error while fetching image ID: %v", err)
+	options := &vpcv1.CreateInstanceOptions{}
+	// Build common field resources, as unique InstancePrototype's are defined based on machine source.
+	profile := &vpcv1.InstanceProfileIdentity{
+		Name: &m.IBMVPCMachine.Spec.Profile,
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply cloud/scope/machine.go:161]: I'll add a safety check, to be replaced by a webhook validation.

File: controllers/ibmvpcmachine_controller.go [hunk 4]
```diff
@@ -162,13 +164,85 @@ func (r *IBMVPCMachineReconciler) reconcileNormal(machineScope *scope.MachineSco
...
+		}
+	} else {
+		machineScope.SetNotReady()
+		conditions.MarkUnknown(machineScope.IBMVPCMachine, infrav1beta2.InstanceReadyCondition, infrav1beta2.InstanceStateUnknownReason, "")
+	}
+
+	// Check if the Machine is ready.
+	if !machineScope.IsReady() {
+		// Requeue after 2 minute if machine is not ready.
+		return ctrl.Result{RequeueAfter: 2 * time.Minute}, nil
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply controllers/ibmvpcmachine_controller.go:225]: I based the time off of what Power expects.

File: controllers/ibmvpcmachine_controller.go [hunk 5]
```diff
@@ -162,13 +164,85 @@ func (r *IBMVPCMachineReconciler) reconcileNormal(machineScope *scope.MachineSco
...
 	if instance != nil {
-		machineScope.IBMVPCMachine.Status.InstanceID = *instance.ID
-		machineScope.IBMVPCMachine.Status.Addresses = []corev1.NodeAddress{
-			{
-				Type:    corev1.NodeInternalIP,
-				Address: *instance.PrimaryNetworkInterface.PrimaryIP.Address,
-			},
+		options := &vpcv1.GetInstanceOptions{}
+		options.SetID(*instance.ID)
+		instanceDetails, _, err := machineScope.IBMVPCClient.GetInstance(options)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply controllers/ibmvpcmachine_controller.go]: Removed.

File: cloud/scope/machine.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjschaef) [reply cloud/scope/machine.go:165]: Added.

File: cloud/scope/machine.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjschaef) [reply cloud/scope/machine.go]: Added.

File: cloud/scope/machine.go [hunk 6]
```diff
@@ -271,6 +450,167 @@ func (m *MachineScope) ensureInstanceUnique(instanceName string) (*vpcv1.Instanc
...
+func (m *MachineScope) getLoadBalancerID(loadBalancer *infrav1beta2.VPCResource) (*string, error) {
+	// Lookup Load Balancer ID by Name if necessary
+	if loadBalancer.ID != nil {
+		return loadBalancer.ID, nil
+	} else if loadBalancer.Name != nil {
+		loadBalancerDetails, err := m.IBMVPCClient.GetLoadBalancerByName(*loadBalancer.Name)
+		if err != nil {
+			return nil, fmt.Errorf("error failed to lookup load balancer id by name %s: %w", *loadBalancer.Name, err)
+		} else if loadBalancerDetails == nil || loadBalancerDetails.ID == nil {
+			return nil, fmt.Errorf("error unable to find laod balancer id with name: %s", *loadBalancer.Name)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply cloud/scope/machine.go]: Fixed.

File: cloud/scope/machine.go [hunk 7]
```diff
@@ -528,6 +943,55 @@ func (m *MachineScope) SetProviderID(id *string) error {
...
+
+	return exists != nil, nil
+}
+
+// TagResource will attach a user Tag to a resource.
+func (m *MachineScope) TagResource(tagName string, resourceCRN string) error {
+	// Verify the Tag to use exists, otherwise create it.
+	exists, err := m.CheckTagExists(tagName)
+	if err != nil {
+		return fmt.Errorf("failure checking if tag exists: %w", err)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply cloud/scope/machine.go]: Added

File: controllers/ibmvpcmachine_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply controllers/ibmvpcmachine_controller.go:218]: In general VPC VM are pretty quick comared to PowerVS so thought  1 min can be sufficient.

[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply controllers/ibmvpcmachine_controller.go:218]: +1

File: cloud/scope/machine.go [hunk 8]
```diff
@@ -100,20 +106,42 @@ func NewMachineScope(params MachineScopeParams) (*MachineScope, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("error failed to create authenticator: %w", err)
+	}
+
+	// Create Global Tagging client.
+	gtOptions := globaltagging.ServiceOptions{
+		GlobalTaggingV1Options: &globaltaggingv1.GlobalTaggingV1Options{
+			Authenticator: auth,
+		},
+	}
```
Comment (Prajyot-Parab) [inline cloud/scope/machine.go:119]: Is it possible to avoid this here and reuse the auth in globaltagging.NewService(gtOptions) and set it there itself?

**Review by Prajyot-Parab — CHANGES_REQUESTED** [review]

File: controllers/ibmvpcmachine_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply controllers/ibmvpcmachine_controller.go:218]: +1

File: cloud/scope/machine.go (hunk 8 above)
Comment (cjschaef) [reply cloud/scope/machine.go]: done

File: controllers/ibmvpcmachine_controller_test.go [hunk 9]
```diff
@@ -347,81 +353,222 @@ func TestIBMVPCMachineLBReconciler_reconcile(t *testing.T) {
...
 			mockvpc.EXPECT().ListLoadBalancerPoolMembers(gomock.AssignableToTypeOf(&vpcv1.ListLoadBalancerPoolMembersOptions{})).Return(&vpcv1.LoadBalancerPoolMemberCollection{}, &core.DetailedResponse{}, nil)
 			mockvpc.EXPECT().CreateLoadBalancerPoolMember(gomock.AssignableToTypeOf(&vpcv1.CreateLoadBalancerPoolMemberOptions{})).Return(customloadBalancerPoolMember, &core.DetailedResponse{}, nil)
-			_, err := reconciler.reconcileNormal(machineScope)
+
+			result, err := reconciler.reconcileNormal(machineScope)
+			// Requeue should be set when the Pool Member is found, but not yet ready (active).
+			g.Expect(result.RequeueAfter).To(Not(BeZero()))
 			g.Expect(err).To(BeNil())
 			g.Expect(machineScope.IBMVPCMachine.Finalizers).To(ContainElement(infrav1beta2.MachineFinalizer))
+			// Machine Status should be ready (running).
```
Comment (Prajyot-Parab) [inline controllers/ibmvpcmachine_controller_test.go:425]: as per comment line 426 should be true? if not can we reword it.

File: controllers/ibmvpcmachine_controller_test.go [hunk 10]
```diff
@@ -347,81 +353,222 @@ func TestIBMVPCMachineLBReconciler_reconcile(t *testing.T) {
...
+			mockgt.EXPECT().AttachTag(gomock.AssignableToTypeOf(&globaltaggingv1.AttachTagOptions{})).Return(nil, &core.DetailedResponse{}, nil)
 			mockvpc.EXPECT().GetLoadBalancer(gomock.AssignableToTypeOf(&vpcv1.GetLoadBalancerOptions{})).Return(loadBalancer, &core.DetailedResponse{}, nil)
 			mockvpc.EXPECT().ListLoadBalancerPoolMembers(gomock.AssignableToTypeOf(&vpcv1.ListLoadBalancerPoolMembersOptions{})).Return(&vpcv1.LoadBalancerPoolMemberCollection{}, &core.DetailedResponse{}, errors.New("failed to list loadBalancerPoolMembers"))
+
 			_, err := reconciler.reconcileNormal(machineScope)
 			g.Expect(err).To(Not(BeNil()))
 			g.Expect(machineScope.IBMVPCMachine.Finalizers).To(ContainElement(infrav1beta2.MachineFinalizer))
 		})
-		t.Run("Should successfully reconcile IBMVPCMachine and set machine status as NotReady when PoolMember is not in active state", func(t *testing.T) {
+		t.Run("Should successfully reconcile IBMVPCMachine and but machine status as not Ready when PoolMember is not in active state yet requiring requeue", func(t *testing.T) {
```
Comment (Prajyot-Parab) [inline controllers/ibmvpcmachine_controller_test.go:404]:
```suggestion
		t.Run("Should successfully reconcile IBMVPCMachine and but machine status as not Ready when PoolMember is not in active state yet requiring requeue", func(t *testing.T) {
```
```suggestion
		t.Run("Should successfully reconcile IBMVPCMachine but its status should be set to Not Ready when the PoolMember is not yet in the active state requiring a requeue", func(t *testing.T) {
```

**Review by Prajyot-Parab — CHANGES_REQUESTED** [review]

File: controllers/ibmvpcmachine_controller_test.go (hunk 10 above)
Comment (Prajyot-Parab) [inline controllers/ibmvpcmachine_controller_test.go:404]:
```suggestion
		t.Run("Should successfully reconcile IBMVPCMachine but its status should be set to Not Ready when the PoolMember is not yet in the active state requiring a requeue", func(t *testing.T) {
```

File: cloud/scope/machine.go (hunk 8 above)
Comment (Prajyot-Parab) [reply cloud/scope/machine.go]: where are the supporting changes in globaltagging service function? Not part of this PR?

File: controllers/ibmvpcmachine_controller_test.go (hunk 10 above)
Comment (cjschaef) [reply controllers/ibmvpcmachine_controller_test.go]: Fixed.

File: controllers/ibmvpcmachine_controller_test.go (hunk 9 above)
Comment (cjschaef) [reply controllers/ibmvpcmachine_controller_test.go]: Updated.

**Comment by mkumatag** [comment]
@Prajyot-Parab can you please review this PR?

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

/hold for @mkumatag approval

File: cloud/scope/machine.go (hunk 8 above)
Comment (Prajyot-Parab) [reply cloud/scope/machine.go]: I hope you tested with above change and it is working fine?

**Comment by mkumatag** [comment]
/unhold

Thanks @Prajyot-Parab @Karthik-K-N

**Pull request merged**